### DISCUSSION TITLE
Render vim.lsp.buf.hover into preview-window

### DISCUSSION THREAD

**Discussion by WhyNotHugo** [neovim/neovim#31670] [Q&A]
By default `vim.lsp.buf.hover()` renders the hover information into a floating window. Is it possible to render it into `preview-window`?

**Comment by WhyNotHugo** [comment]
Looks like `M.hover` in  `runtime/lua/vim/lsp/handlers.lua` hardcodes `util.open_floating_preview`.

I guess a new flag would be needed to customise this behaviour.

**Reply by clason** [reply]
Nope, out of the question. Just write your own handler, it's not hard.

**Reply by WhyNotHugo** [reply]
Actually, the above handler is no longer used (it's marked deprecated).

The call for hover happens in `runtime/lua/vim/lsp/buf.lua`, inside `M.hover`. The logic to execute the call too all LSPs, prompt for a single one and render the floating window is all intertwined in one function, so the entire functionality would have to be rewritten.

**Reply by WhyNotHugo** [reply]
> Nope, out of the question.

Can you elaborate on this a bit? Displaying the result of `hover` in a `preview-window` would be immensely useful.

> Just write your own handler, it's not hard.

Writing my own handler wouldn't work, because the handler is not called. A handler function is hardcoded inline inside `vim.lsp.buf.hover()`

**Comment by WhyNotHugo** [comment]
For anyone else who's interested in being able to read lsp-hover information, here's a snippet which uses preview-window. It's somewhat of a hack, but it works, and this is exactly the intended purpose of preview-window.

```lua
-- Show keyword or lsp-hover in preview-window.
--
-- Overrides the default behaviour (a pop-up) which is useless for reading documentation.
--
-- Ignores the following variables, which are never defined for my use cases:
--   opts.close_events, opts.focusable, opts.focus, opts.height
---@diagnostic disable-next-line: duplicate-set-field -- Overwriting field.
vim.lsp.util.open_floating_preview = function(contents, syntax, opts)
  -- Hint: use ctrl-w_p to focus preview window.
  -- Hint: when focused, esc closes the preview window.
  -- Hint: use :pclose to close the preview window without focusing it.

  local orig_win = vim.api.nvim_get_current_win()

  vim.api.nvim_command("pedit")
  vim.api.nvim_command("wincmd P")

  local hover_bufnr = vim.api.nvim_create_buf(false, true)
  vim.api.nvim_set_current_buf(hover_bufnr)
  vim.api.nvim_buf_set_lines(hover_bufnr, 0, -1, false, contents)
  vim.api.nvim_buf_set_name(hover_bufnr, opts.focus_id or "LSP Hover")
  vim.api.nvim_set_option_value("filetype", "lsp_hover", { buf = hover_bufnr })
  -- Overwrites previous filetype, but ftplugin executed already anyway.
  vim.api.nvim_set_option_value("filetype", syntax, { buf = hover_bufnr })

  vim.api.nvim_set_current_win(orig_win)
end
```

Additionally, `ftplugin/lsp_hover.vim` contains:

```vim
setlocal bufhidden=wipe
setlocal nomodifiable
setlocal conceallevel=3
setlocal concealcursor=nvic
setlocal nospell
" Close with Esc.
nnoremap <buffer> <Esc> :q<CR>
```
